### DISCUSSION TITLE
v2.7.0

### DISCUSSION THREAD

**Discussion by epwalsh** [allenai/allennlp#5394] [General]
### Added 🎉

- Added support to evaluate mutiple datasets and produce corresponding output files in the `evaluate` command.
- Added more documentation to the learning rate schedulers to include a sample config object for how to use it.
- Moved the pytorch learning rate schedulers wrappers to their own file called `pytorch_lr_schedulers.py` so that they will have their own documentation page.
- Added a module `allennlp.nn.parallel` with a new base class, `DdpAccelerator`, which generalizes
  PyTorch's `DistributedDataParallel` wrapper to support other implementations. Two implementations of
  this class are provided. The default is `TorchDdpAccelerator` (registered at "torch"), which is just a thin wrapper around
  `DistributedDataParallel`. The other is `FairScaleFsdpAccelerator`, which wraps FairScale's
  [`FullyShardedDataParallel`](https://fairscale.readthedocs.io/en/latest/api/nn/fsdp.html).
  You can specify the `DdpAccelerator` in the "distributed" section of a configuration file under the key "ddp_accelerator".
- Added a module `allennlp.nn.checkpoint` with a new base class, `CheckpointWrapper`, for implementations
  of activation/gradient checkpointing. Two implentations are provided. The default implementation is `TorchCheckpointWrapper` (registered as "torch"),
  which exposes [PyTorch's checkpoint functionality](https://pytorch.org/docs/stable/checkpoint.html).
  The other is `FairScaleCheckpointWrapper` which exposes the more flexible
  [checkpointing funtionality from FairScale](https://fairscale.readthedocs.io/en/latest/api/nn/checkpoint/checkpoint_activations.html).
- The `Model` base class now takes a `ddp_accelerator` parameter (an instance of `DdpAccelerator`) which will be available as
  `self.ddp_accelerator` during distributed training. This is useful when, for example, instantiating submodules in your
  model's `__init__()` method by wrapping them with `self.ddp_accelerator.wrap_module()`. See the `allennlp.modules.transformer.t5`
  for an example.
- We now log batch metrics to tensorboard and wandb.
- Added Tango components, to be explored in detail in a later post
- Added `ScaledDotProductMatrixAttention`, and converted the transformer toolkit to use it
- Added tests to ensure that all `Attention` and `MatrixAttention` implementations are interchangeable
- Added a way for AllenNLP Tango to read and write datasets lazily.
- Added a way to remix datasets flexibly
- Added `from_pretrained_transformer_and_instances` constructor to `Vocabulary`
- `TransformerTextField` now supports `__len__`.

### Fixed ✅

- Fixed a bug in `ConditionalRandomField`: `transitions` and `tag_sequence` tensors were not initialized on the desired device causing high CPU usage (see https://github.com/allenai/allennlp/issues/2884)
- Fixed a mispelling: the parameter `contructor_extras` in `Lazy()` is now correctly called `constructor_extras`.
- Fixed broken links in `allennlp.nn.initializers` docs.
- Fixed bug in `BeamSearch` where `last_backpointers` was not being passed to any `Constraint`s.
- `TransformerTextField` can now take tensors of shape `(1, n)` like the tensors produced from a HuggingFace tokenizer.
- `tqdm` lock is now set inside `MultiProcessDataLoading` when new workers are spawned to avoid contention when writing output.
- `ConfigurationError` is now pickleable.
- Checkpointer cleaning was fixed to work on Windows Paths
- Multitask models now support `TextFieldTensor` in heads, not just in the backbone.
- Fixed the signature of `ScaledDotProductAttention` to match the other `Attention` classes
- `allennlp` commands will now catch `SIGTERM` signals and handle them similar to `SIGINT` (keyboard interrupt).
- The `MultiProcessDataLoader` will properly shutdown its workers when a `SIGTERM` is received.
- Fixed the way names are applied to Tango `Step` instances.
- Fixed a bug in calculating loss in the distributed setting.
- Fixed a bug when extending a sparse sequence by 0 items.

### Changed ⚠️

- The type of the `grad_norm` parameter of `GradientDescentTrainer` is now `Union[float, bool]`,
  with a default value of `False`. `False` means gradients are not rescaled and the gradient
  norm is never even calculated. `True` means the gradients are still not rescaled but the gradient
  norm is calculated and passed on to callbacks. A `float` value means gradients are rescaled.
- `TensorCache` now supports more concurrent readers and writers.
- We no longer log parameter statistics to tensorboard or wandb by default.

## Commits

48af9d34 Multiple datasets and output files support for the evaluate command (#5340)
60213cd7 Tiny tango tweaks (#5383)
28950215 improve signal handling and worker cleanup (#5378)
b41cb3eb Fix distributed loss (#5381)
6355f073 Fix Checkpointer cleaner regex on Windows (#5361)
27da04cf Dataset remix (#5372)
75af38e0 Create Vocabulary from both pretrained transformers and instances (#5368)
5dc80a65 Adds a dataset that can be read and written lazily (#5344)
01e8a35a Improved Documentation For Learning Rate Schedulers (#5365)
8370cfa3 skip loading t5-base in CI (#5371)
13de38d1 Log batch metrics (#5362)
1f5c6e5b Use our own base images to build allennlp Docker images (#5366)
bffdbfd1 Bugfix: initializing all tensors and parameters of the `ConditionalRandomField` model on the proper device (#5335)
d45a2dab Make sure that all attention works the same (#5360)
c1edaef8 Update google-cloud-storage requirement (#5357)
524244b6 Update wandb requirement from <0.12.0,>=0.10.0 to >=0.10.0,<0.13.0 (#5356)
90bf33b8 small fixes for tango (#5350)
2e11a15e tick version for nightly releases
311f1104 Tango (#5162)
1df2e517 Bump fairscale from 0.3.8 to 0.3.9 (#5337)
b72bbfc9 fix constraint bug in beam search, clean up tests (#5328)
ec3e2943 Create CITATION.cff (#5336)
8714aa0b This is a desperate attempt to make TensorCache a little more stable (#5334)
fd429b2b Update transformers requirement from <4.9,>=4.1 to >=4.1,<4.10 (#5326)
1b5ef3a0 Update spacy requirement from <3.1,>=2.1.0 to >=2.1.0,<3.2 (#5305)
1f20513d TextFieldTensor in multitask models (#5331)
76f2487b set tqdm lock when new workers are spawned (#5330)
67add9d9 Fix `ConfigurationError` deserialization (#5319)
42d85298 allow TransformerTextField to take input directly from HF tokenizer (#5329)
64043ac6 Bump black from 21.6b0 to 21.7b0 (#5320)
32750550 Update mkdocs-material requirement from <7.2.0,>=5.5.0 to >=5.5.0,<7.3.0 (#5327)
5b1da908 Update links in initializers documentation (#5317)
ca656fc6 FairScale integration (#5242)

<hr /><em>This discussion was created from the release <a href='https://github.com/allenai/allennlp/releases/tag/v2.7.0'>v2.7.0</a>.</em>
